stan/math/rev/fun/inverse.hpp

```diff
@@ -75,16 +21,23 @@ class inverse_vari : public vari {
  * size zero).
  * @throw std::invalid_argument if the matrix is not square.
  */
-inline matrix_v inverse(const matrix_v &m) {
+template <typename T, require_eigen_vt<is_var, T>* = nullptr>
+inline Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic> inverse(const T& m) {
   check_square("inverse", "m", m);
   if (m.size() == 0) {
     return {};
   }
 
-  matrix_v res(m.rows(), m.cols());
-  internal::inverse_vari *baseVari = new internal::inverse_vari(m);
-  res.vi() = Eigen::Map<matrix_vi>(baseVari->vari_ref_A_inv_, res.rows(),
-                                   res.cols());
+  arena_matrix<Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>> arena_m = m;
+  arena_matrix<Eigen::MatrixXd> res_val = value_of(arena_m).inverse();
+  arena_matrix<Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>> res
+      = res_val;
+
+  reverse_pass_callback([res, res_val, arena_m]() mutable {
+    Eigen::MatrixXd res_adj = res.adj();
+    arena_m.adj() -= res_val.transpose() * res_adj * res_val.transpose();
```

Comment:
This code didn't compile and I don't know why:

```
arena_m.adj() -= res_val.transpose() * res.adj() * res_val.transpose();
```

Reply:
When calling ```.val()```/```.adj()``` on a non-const Eigen matrix the returned type is a ```CWiseUnaryView```, and there's currently a bug with multiplying two ```CWiseUnaryView``` objects ([reported here](https://gitlab.com/libeigen/eigen/-/issues/1730)). I think that might be the source of your problems

Reply:
Cool beans, I'll just leave it as is then, thanks